### PR TITLE
Use new clients in lambda event source mappings

### PR DESCRIPTION
**Pull request by dfangl**
## Motivation
To provide IAM enforcement, we have to use the new clients, with the execution role assumed, for the event source mapping operations.

## Changes
* Use Adapters to get the right clients for the operation
* Migrate to new client factory
* Consolidate `wait_for_dynamodb_stream_ready` fixtures
* Handle failure for get_records operation in the streams event source listener separately, to avoid crashing the whole thread

### CODE REVIEW

File: localstack/testing/pytest/fixtures.py [hunk 1]
```diff
@@ -656,7 +656,7 @@ def is_stream_ready():
                 "UPDATING",
             ]
```
Comment (joe4dev) [inline localstack/testing/pytest/fixtures.py:658]: idea: The triple nesting can be confusing to use. A usage example would be helpful.

File: localstack/services/awslambda/event_source_listeners/adapters.py [hunk 2]
```diff
@@ -258,3 +270,29 @@ def get_event_sources(self, source_arn: str):
...
+
+    def _get_role_for_function(self, function_arn: str) -> str:
+        function_name, qualifier, account, region = function_locators_from_arn(function_arn)
+        store = lambda_stores[account][region]
+        function = store.functions.get(function_name)
+        if qualifier and qualifier != "$LATEST":
+            if qualifier_is_version(qualifier):
+                version_number = qualifier
+            else:
+                # the role of the routing config version and the regular configured version has to be identical
```
Comment (joe4dev) [inline localstack/services/awslambda/event_source_listeners/adapters.py:288]: do we have snapshot tests for this?

File: localstack/services/awslambda/event_source_listeners/kinesis_event_source_listener.py [hunk 3]
```diff
@@ -64,9 +66,7 @@ def _create_lambda_event_payload(self, stream_arn, records, shard_id=None):
                     "eventSource": "aws:kinesis",
                     "eventVersion": "1.0",
                     "eventName": "aws:kinesis:record",
-                    "invokeIdentityArn": "arn:aws:iam::{0}:role/lambda-role".format(
-                        get_aws_account_id()
-                    ),
+                    "invokeIdentityArn": f"arn:aws:iam::{get_aws_account_id()}:role/lambda-role",
```
Comment (joe4dev) [inline localstack/services/awslambda/event_source_listeners/kinesis_event_source_listener.py:69]: ✨ 👍

File: localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py [hunk 4]
```diff
@@ -198,63 +200,75 @@ def _listen_to_shard_and_invoke_lambda(self, params: Dict):
...
+                                        failure_destination,
+                                    )
+                            else:
+                                should_get_next_batch = False
+                    if should_get_next_batch:
+                        shard_iterator = records_response["NextShardIterator"]
+                        num_invocation_failures = 0
                 time.sleep(self._POLL_INTERVAL_SEC)
         except Exception as e:
+            #
```
Comment (joe4dev) [inline localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py:271]: comment missing or typo ?

File: localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py [hunk 5]
```diff
@@ -198,63 +200,75 @@ def _listen_to_shard_and_invoke_lambda(self, params: Dict):
...
+                try:
+                    records_response = stream_client.get_records(
+                        ShardIterator=shard_iterator, Limit=batch_size
                     )
-                    if is_invocation_successful:
-                        should_get_next_batch = True
+                except ClientError as e:
+                    if "AccessDeniedException" in str(e):
+                        LOG.warning(
+                            "Insufficient permissions to get records from stream %s: %s",
```
Comment (joe4dev) [inline localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py:210]: Are these types of IAM exceptions logged somewhere else on AWS?

File: localstack/testing/pytest/fixtures.py (hunk 1 above)
Comment (dfangl) [reply localstack/testing/pytest/fixtures.py:658]: You are right but that is pretty unrelated to the PR. Would definitely be nice to add more docs to the test fixtures though!

File: localstack/testing/pytest/fixtures.py [hunk 6]
```diff
@@ -656,7 +656,7 @@ def is_stream_ready():
...
 
-        poll_condition(is_stream_ready)
+        return poll_condition(is_stream_ready)
```
[the remark this answers is not in the document]
Comment (dfangl) [reply localstack/testing/pytest/fixtures.py:659]: The return is needed if you do something like:
```
assert wait_for_stream_ready()
```
So you can make sure that it not only waited for it, but the wait also ended successfully (not just a timeout). I merged another fixture with this one, which is why this is needed.
We should do it everywhere, but the user of the fixture also has to check the return value for this, which would make this a bigger adventure.

File: localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py (hunk 4 above)
Comment (dfangl) [reply localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py:271]: Whoopsie, that's not supposed to be in there. Will remove!

File: localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py (hunk 5 above)
Comment (dfangl) [reply localstack/services/awslambda/event_source_listeners/stream_event_source_listener.py:210]: Might be in cloudtrail, but as it is a data operation, probably not even there. So no, it's very hard in AWS to detect if something like this happens.

This is why in AWS you have a test for the permissions before you create the event source mappings, but that is kind of lambda specific. We do not have that yet, but it is definitely planned to be added.

File: localstack/services/awslambda/event_source_listeners/adapters.py (hunk 2 above)
Comment (dfangl) [reply localstack/services/awslambda/event_source_listeners/adapters.py:288]: I do not think so, no.

**Pull request merged**